src/overlays/actors/ovl_Arms_Hook/z_arms_hook.c

```diff
@@ -23,27 +29,329 @@ const ActorInit Arms_Hook_InitVars = {
+    if (func_8008F104(player) != 0) {
+        if ((player->unk_154 != player->unk_151) || ((player->actor.flags & 0x100) != 0) ||
+            ((player->stateFlags1 & 0x4000080) != 0)) {
+            this->unk_210 = 0;
+            ArmsHook_DetachHookFromActor(this);
+            Math_Vec3f_Copy(&this->actor.posRot.pos, &player->unk_3C8);
+            return 1;
+        }
+    }
+    return 0;
+}
+
+void ArmsHook_AttachHookToActor(ArmsHook* this, Actor* actor) {
+    actor->flags |= FLAG_HOOKSHOT_HOOKED;
+    this->hookedActor = actor;
+    Math_Vec3f_Diff(&actor->posRot.pos, &this->actor.posRot.pos, &this->hookedActorDistDiff);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Arms_Hook/ArmsHook_Destroy.s")
+void func_8086518C(ArmsHook* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+    Actor* at;
+    Actor* hookedActor;
+    Vec3f hookBodyDistDiffVec;
+    Vec3f newPos;
+    f32 hookBodyDistDiff;
+    f32 phi_f16;
+    DynaPolyActor* dynaPolyActor;
+    f32 sp94;
+    f32 sp90;
+    s16 phi_v0;
+    CollisionPoly* poly;
+    u32 dynaPolyID;
+    Vec3f sp78;
+    Vec3f prevFrameDiff;
+    Vec3f sp60;
+    f32 sp5C;
+    f32 sp58;
+    f32 velocity;
+    Vec3f* hookBodyDistDiffVecPtr;
+    Vec3f* curPos;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Arms_Hook/func_80864FC4.s")
+    if ((this->actor.attachedA == 0) || (func_8008F104(player) == 0)) {
+        ArmsHook_DetachHookFromActor(this);
+        Actor_Kill(&this->actor);
+        return;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Arms_Hook/func_80865044.s")
+    curPos = &this->actor.posRot.pos;
+    func_8002F8F0(&player->actor, 0x100B);
+    func_808650AC(this);
+    if (this->unk_210 != 0) {
+        if ((this->colliderQuad.base.atFlags & 2) != 0) {
+            if (this->colliderQuad.body.atHitItem->flags != 4) {
+                at = this->colliderQuad.base.at;
+                if (at->update != 0) {
+                    if ((at->flags & 0x600) != 0) {
+                        if (this->colliderQuad.body.atHitItem->bumperFlags & 4) {
+                            ArmsHook_AttachHookToActor(this, at);
+                            if ((at->flags & 0x400) == 0x400) {
+                                func_80865044(this);
+                            }
+                        }
+                    }
+                }
+                this->unk_210 = 0;
+                Audio_PlaySoundGeneral(NA_SE_IT_ARROW_STICK_CRE, &this->actor.unk_E4, 4, &D_801333E0, &D_801333E0,
+                                       &D_801333E8);
+                return;
+            }
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Arms_Hook/func_80865054.s")
+    if (this->unk_210 == 0) {
+        phi_v0 = 0;
+    } else {
+        phi_v0 = --this->unk_210;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Arms_Hook/func_80865084.s")
+    if (phi_v0 == 0) {
```

Comment:
The whole `phi_v0` stuff before can most likely be replaced with an usage of the `DECR` macro here.
```suggestion
    if (DECR(this->unk_210) == 0) {
```